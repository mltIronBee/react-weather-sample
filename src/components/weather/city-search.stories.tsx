import React, { ComponentProps } from "react";
import { Story } from "@storybook/react/types-6-0";
import { CitySearch } from "@components/weather";

const Template: Story<ComponentProps<typeof CitySearch>> = (args) => <CitySearch {...args} />;

export const Default = Template.bind({});

Default.args = {
	value: "Odesa",
};

export const Loading = Template.bind({});

Loading.args = {
	value: "Odesa",
	searching: true,
};

export const WithError = Template.bind({});

WithError.args = {
	value: "Lorem Ipsum",
	errorMessage: "Search error",
};

export default {
	title: "CitySearch",
	component: CitySearch,
};
