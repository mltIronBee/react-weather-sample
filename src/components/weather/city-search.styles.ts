import { Theme } from "@material-ui/core";
import { makeStyles, createStyles } from "@material-ui/styles";

export default makeStyles((theme: Theme) =>
	createStyles({
		loadingIcon: {
			padding: theme.spacing() * 1.5,
		},
	}),
);
