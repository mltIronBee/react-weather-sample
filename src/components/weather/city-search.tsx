import React from "react";
import TextField from "@material-ui/core/TextField";
import InputAdornment from "@material-ui/core/InputAdornment";
import IconButton from "@material-ui/core/IconButton";
import SearchIcon from "@material-ui/icons/Search";
import CircularProgress from "@material-ui/core/CircularProgress";
import useStyles from "@components/weather/city-search.styles";

export interface ICitySearchProps {
	value: string;
	searching: boolean;
	errorMessage?: string;
	onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
	onSearch: (event: React.MouseEvent<HTMLButtonElement>) => void | Promise<void>;
}

export const CitySearch: React.FC<ICitySearchProps> = (props) => {
	const classes = useStyles();

	return (
		<TextField
			label="Search city"
			error={!!props.errorMessage}
			helperText={props.errorMessage}
			value={props.value}
			onChange={props.onChange}
			variant="outlined"
			InputProps={{
				endAdornment: (
					<InputAdornment position="end">
						{props.searching ? (
							<CircularProgress className={classes.loadingIcon} size={24} />
						) : (
							<IconButton onClick={props.onSearch}>
								<SearchIcon />
							</IconButton>
						)}
					</InputAdornment>
				),
			}}
		/>
	);
};
