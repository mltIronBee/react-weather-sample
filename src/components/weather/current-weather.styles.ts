import { makeStyles } from "@material-ui/core/styles";

export default makeStyles({
	sunriseSunsetContainer: {
		width: "100%",
		display: "flex",
		justifyContent: "space-evenly",
	},
	sunriseIcon: {
		color: "#EC6E4C",
	},
	sunsetIcon: {
		color: "#EC6E4C",
	},
	weatherContainer: {
		display: "flex",
		justifyContent: "center",
		alignItems: "center",
	},
	weatherDescription: {
		display: "inline-block",
		textTransform: "capitalize",
	},
	weatherIcon: {
		display: "inline-block",
	},
	windDirectionIcon: (windDeg: number) => ({
		transform: `rotate(${windDeg}deg)`,
	}),
});
