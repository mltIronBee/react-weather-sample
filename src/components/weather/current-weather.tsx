import React, { memo } from "react";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import Divider from "@material-ui/core/Divider";
import useStyles from "@components/weather/current-weather.styles";
import ListItemIcon from "@material-ui/core/ListItemIcon";
import DownIcon from "@material-ui/icons/VerticalAlignBottom";
import CloudIcon from "@material-ui/icons/Cloud";
import DropIcon from "@material-ui/icons/Opacity";
import SunIcon from "@material-ui/icons/WbSunny";
import SnowIcon from "@material-ui/icons/AcUnit";
import VisibilityIcon from "@material-ui/icons/Visibility";
import SpeedIcon from "@material-ui/icons/Speed";
import ArrowUpwardIcon from "@material-ui/icons/ArrowUpward";
import RainIcon from "@components/icons/rain";
import SunriseIcon from "@components/icons/sunrise";
import SunsetIcon from "@components/icons/sunset";

export interface ICurrentWeatherProps {
	date: string;
	sunrise: string;
	sunset: string;
	temperature: number;
	feelsLike: number;
	pressure: number;
	humidity: number;
	clouds: number;
	uvIndex: number;
	visibility: number;
	rain?: number;
	snow?: number;
	windSpeed: number;
	windDeg: number;
	weatherIcon: string;
	weatherDescription: string;
}

const formatTemperature = (temperature: number): string =>
	temperature > 0 ? `+${Math.round(temperature)}` : Math.round(temperature).toString(10);

const getReadableWindDirection = (windDeg: number): string => {
	if (windDeg > 337.5 || windDeg <= 22.5) {
		return "North";
	} else if (windDeg > 22.5 && windDeg <= 67.5) {
		return "North-East";
	} else if (windDeg > 67.5 && windDeg <= 112.5) {
		return "East";
	} else if (windDeg > 112.5 && windDeg <= 157.5) {
		return "South-East";
	} else if (windDeg > 157.5 && windDeg <= 202.5) {
		return "South";
	} else if (windDeg > 202.5 && windDeg <= 247.5) {
		return "South-West";
	} else if (windDeg > 247.5 && windDeg <= 292.5) {
		return "West";
	}

	return "North-West";
};

export const CurrentWeather: React.FC<ICurrentWeatherProps> = memo((props) => {
	const classes = useStyles(props.windDeg);

	return (
		<Grid container>
			<Grid item xs={12} md={5} lg={4} xl={3}>
				<Grid container direction="column" justify="center" alignItems="center">
					<Typography>{props.date}</Typography>
					<span className={classes.weatherContainer}>
						<img className={classes.weatherIcon} alt={props.weatherDescription} src={props.weatherIcon} />
						<Typography className={classes.weatherDescription} variant="h4">
							{props.weatherDescription}
						</Typography>
					</span>
					<Typography gutterBottom variant="h2">
						{formatTemperature(props.temperature)}&deg;
					</Typography>
					<Typography gutterBottom variant="h4">
						Feels like: {formatTemperature(props.feelsLike)}&deg;
					</Typography>
					<Typography gutterBottom component="span" className={classes.sunriseSunsetContainer}>
						<Typography align="center" variant="h5" display="inline">
							<SunriseIcon className={classes.sunriseIcon} /> Sunrise at {props.sunrise}
						</Typography>
						<Typography align="center" variant="h5" display="inline">
							<SunsetIcon className={classes.sunsetIcon} /> Sunset at {props.sunset}
						</Typography>
					</Typography>
				</Grid>
			</Grid>
			<Grid item xs={12} md={7} lg={8} xl={9}>
				<List>
					<ListItem>
						<ListItemIcon>
							<DownIcon />
						</ListItemIcon>
						<ListItemText primary={`${props.pressure} hPa`} secondary="Atmospheric pressure" />
					</ListItem>
					<Divider />
					<ListItem>
						<ListItemIcon>
							<DropIcon />
						</ListItemIcon>
						<ListItemText primary={`${props.humidity} %`} secondary="Humidity" />
					</ListItem>
					<Divider />
					<ListItem>
						<ListItemIcon>
							<CloudIcon />
						</ListItemIcon>
						<ListItemText primary={`${props.clouds} %`} secondary="Clouds" />
					</ListItem>
					<Divider />
					<ListItem>
						<ListItemIcon>
							<SunIcon />
						</ListItemIcon>
						<ListItemText primary={props.uvIndex} secondary="Midday UV Index" />
					</ListItem>
					<Divider />
					<ListItem>
						<ListItemIcon>
							<VisibilityIcon />
						</ListItemIcon>
						<ListItemText primary={`${props.visibility} m`} secondary="Visibility" />
					</ListItem>
					<Divider />
					{props.rain && (
						<React.Fragment>
							<ListItem>
								<ListItemIcon>
									<RainIcon />
								</ListItemIcon>
								<ListItemText primary={`${props.rain} mm`} secondary="Rain last 1 hour" />
							</ListItem>
							<Divider />
						</React.Fragment>
					)}
					{props.snow && (
						<React.Fragment>
							<ListItem>
								<ListItemIcon>
									<SnowIcon />
								</ListItemIcon>
								<ListItemText primary={`${props.snow} mm`} secondary="Snow last 1 hour" />
							</ListItem>
							<Divider />
						</React.Fragment>
					)}
					<ListItem>
						<ListItemIcon>
							<SpeedIcon />
						</ListItemIcon>
						<ListItemText primary={`${props.windSpeed} m/s`} secondary="Wind speed" />
					</ListItem>
					<Divider />
					<ListItem>
						<ListItemIcon>
							<ArrowUpwardIcon className={classes.windDirectionIcon} />
						</ListItemIcon>
						<ListItemText primary={getReadableWindDirection(props.windDeg)} secondary="Wind direction" />
					</ListItem>
				</List>
			</Grid>
		</Grid>
	);
});

CurrentWeather.displayName = "CurrentWeather";
