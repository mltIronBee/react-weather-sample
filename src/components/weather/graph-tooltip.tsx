import React from "react";
import Paper from "@material-ui/core/Paper";
import Typography from "@material-ui/core/Typography";
import { TooltipProps } from "recharts";
import { Divider } from "@material-ui/core";
import useStyles from "@components/weather/graph-tooltip.styles";
import { useTranslation } from "react-i18next";

export const GraphTooltip: React.FC<TooltipProps> = (props) => {
	const classes = useStyles();
	const { t } = useTranslation(["forecast"]);

	if (!props.active || !props.payload) {
		return null;
	}

	return (
		<Paper className={classes.container} elevation={4}>
			<Typography variant="h6">{props.label}</Typography>
			<Divider className={classes.divider} />
			<Typography className={classes.minTemperatureLabel}>
				{t("forecast:tooltip-min")}: {props.payload[0].value > 0 ? "+" : ""}
				{Math.round(+props.payload[0].value)}
			</Typography>
			<Typography className={classes.maxTemperatureLabel}>
				{t("forecast:tooltip-max")}: {props.payload[1].value > 0 ? "+" : ""}
				{Math.round(+props.payload[1].value)}
			</Typography>
		</Paper>
	);
};
