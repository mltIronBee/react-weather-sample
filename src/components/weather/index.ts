export * from "@components/weather/city-search";
export * from "@components/weather/forecast-graph";
export * from "@components/weather/tab-controls";
export * from "@components/weather/current-weather";
