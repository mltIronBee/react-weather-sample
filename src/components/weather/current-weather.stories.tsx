import React, { ComponentProps, useState } from "react";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import Slider from "@material-ui/core/Slider";
import { DateTime } from "luxon";
import { Story } from "@storybook/react/types-6-0";
import { CurrentWeather, ICurrentWeatherProps } from "@components/weather";
import { getWeatherIcon } from "@utils/weather-api";

const WindDirectionWrapper: React.FC<Omit<ICurrentWeatherProps, "windDeg">> = (props) => {
	const [windDeg, setWindDeg] = useState(0);

	const handleWindDirectionChange = (event: React.ChangeEvent<unknown>, value: number | number[]): void => {
		setWindDeg(value as number);
	};

	return (
		<Grid container>
			<Typography gutterBottom>Wind direction</Typography>
			<Grid spacing={3} container>
				<Grid item>
					<Typography>0</Typography>
				</Grid>
				<Grid item xs>
					<Slider min={0} max={360} value={windDeg} onChange={handleWindDirectionChange} />
				</Grid>
				<Grid item>
					<Typography>360</Typography>
				</Grid>
			</Grid>
			<Grid container>
				<CurrentWeather windDeg={windDeg} {...props} />
			</Grid>
		</Grid>
	);
};

const Template: Story<ComponentProps<typeof WindDirectionWrapper>> = (args) => <WindDirectionWrapper {...args} />;

export const Default = Template.bind({});

Default.args = {
	date: DateTime.local().toLocaleString(DateTime.DATETIME_FULL),
	sunrise: "06:00",
	sunset: "18:00",
	temperature: 24,
	feelsLike: 20,
	pressure: 1000,
	humidity: 80,
	clouds: 0,
	uvIndex: 0,
	visibility: 10000,
	windSpeed: 5,
	weatherIcon: getWeatherIcon("01d"),
	weatherDescription: "Clear sky",
};

export const WithRain = Template.bind({});

WithRain.args = {
	date: DateTime.local().toLocaleString(DateTime.DATETIME_FULL),
	sunrise: "06:00",
	sunset: "18:00",
	temperature: 5,
	feelsLike: -3,
	pressure: 1000,
	humidity: 80,
	clouds: 0,
	uvIndex: 0,
	visibility: 10000,
	windSpeed: 5,
	weatherIcon: getWeatherIcon("09d"),
	weatherDescription: "shower rain",
	rain: 30,
};

export const WithSnow = Template.bind({});

WithSnow.args = {
	date: DateTime.local().toLocaleString(DateTime.DATETIME_FULL),
	sunrise: "06:00",
	sunset: "18:00",
	temperature: -14,
	feelsLike: -20,
	pressure: 1000,
	humidity: 80,
	clouds: 0,
	uvIndex: 0,
	visibility: 10000,
	windSpeed: 5,
	weatherIcon: getWeatherIcon("13d"),
	weatherDescription: "Snow",
	snow: 30,
};

export default {
	title: "Current weather component",
	component: CurrentWeather,
};
