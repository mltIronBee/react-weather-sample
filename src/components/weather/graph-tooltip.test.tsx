import React from "react";
import { render } from "@utils/test-utils";
import { GraphTooltip } from "@components/weather/graph-tooltip";
import "@testing-library/jest-dom/extend-expect";

describe("Graph tooltip", () => {
	describe("Rendering", () => {
		it("Should render correctly", () => {
			const { container } = render(
				<GraphTooltip
					active
					label="Test label"
					payload={[
						{ name: "Test name 1", value: 10 },
						{ name: "Test name 2", value: -10 },
					]}
				/>,
			);

			expect(container.firstChild).toMatchSnapshot();
		});
	});
});
