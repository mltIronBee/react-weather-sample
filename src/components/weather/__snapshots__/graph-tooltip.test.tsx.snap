// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph tooltip Rendering Should render correctly 1`] = `
<div
  class="MuiPaper-root makeStyles-container MuiPaper-elevation4 MuiPaper-rounded"
>
  <h6
    class="MuiTypography-root MuiTypography-h6"
  >
    Test label
  </h6>
  <hr
    class="MuiDivider-root makeStyles-divider"
  />
  <p
    class="MuiTypography-root makeStyles-minTemperatureLabel MuiTypography-body1"
  >
    Min Temperature
    : 
    +
    10
  </p>
  <p
    class="MuiTypography-root makeStyles-maxTemperatureLabel MuiTypography-body1"
  >
    Max Temperature
    : 
    
    -10
  </p>
</div>
`;
